ext/exif/exif.c

```diff
@@ -4285,11 +4297,125 @@ static bool exif_process_IFD_in_TIFF(image_info_type *ImageInfo, size_t dir_offs
 	return result;
 }
 
+static int exif_isobmff_parse_box(unsigned char *buf, isobmff_box_type *box)
+{
+	box->size = php_ifd_get32u(buf, 1);
+	buf += 4;
+	box->type = php_ifd_get32u(buf, 1);
+	if (box->size != 1) {
+		return 8;
+	}
+	buf += 4;
+	box->size = php_ifd_get64u(buf, 1);
+	return 16;
+}
+
+static void exif_isobmff_parse_meta(unsigned char *data, unsigned char *end, isobmff_item_pos_type *pos)
+{
+	isobmff_box_type box, item;
+	unsigned char *box_offset, *p, *p2;
+	int header_size, exif_id = -1, version, item_count, i;
+
+	for (box_offset = data + 4; box_offset < end; box_offset += box.size) {
+		header_size = exif_isobmff_parse_box(box_offset, &box);
+		if (box.type == FOURCC("iinf")) {
+			p = box_offset + header_size;
+			version = p[0];
+			p += 4;
+			if (version < 2) {
+				item_count = php_ifd_get16u(p, 1);
+				p += 2;
+			} else {
+				item_count = php_ifd_get32u(p, 1);
+				p += 4;
+			}
+			for (i=0; i<item_count; i++) {
+				header_size = exif_isobmff_parse_box(p, &item);
+				if (!memcmp(p + header_size + 8, "Exif", 4)) {
+					exif_id = php_ifd_get16u(p + header_size + 4, 1);
+					break;
+				}
+				p += item.size;
+			}
+			if (exif_id < 0) {
+				break;
+			}
+		}
+		else if (box.type == FOURCC("iloc")) {
+			p = box_offset + header_size;
+			version = p[0];
+			p += 6;
+			if (version < 2) {
+				item_count = php_ifd_get16u(p, 1);
+				p += 2;
+			} else {
+				item_count = php_ifd_get32u(p, 1);
+				p += 4;
+			}
+			for (i=0, p2=p; i<item_count; i++, p2 += 16) {
+				fflush(stdout);
+				if (php_ifd_get16u(p2, 1) == exif_id) {
+					pos->offset = php_ifd_get32u(p2 + 8, 1);
+					pos->size = php_ifd_get32u(p2 + 12, 1);
+					break;
+				}
+			}
+			break;
+		}
+	}
+}
+
+static bool exif_scan_HEIF_header(image_info_type *ImageInfo, unsigned char *buf)
+{
+	isobmff_box_type box;
+	isobmff_item_pos_type pos;
+	unsigned char *data;
+	off_t offset;
+	uint64_t limit;
+	int box_header_size, remain;
+	bool ret = false;
+
+	pos.size = 0;
+	for (offset = php_ifd_get32u(buf, 1); ImageInfo->FileSize > offset + 16; offset += box.size) {
+		if ((php_stream_seek(ImageInfo->infile, offset, SEEK_SET) < 0) ||
+			(exif_read_from_stream_file_looped(ImageInfo->infile, (char*)buf, 16) != 16)) {
+			break;
+		}
+		box_header_size = exif_isobmff_parse_box(buf, &box);
+		if (box.type == FOURCC("meta")) {
+			limit = box.size - box_header_size;
+			data = (unsigned char *)safe_emalloc(1, limit, 0);
```

Comment:
I don't think so. 
The data in buffer is useless, if we call erealloc and the new size is bigger, the OS had to copy all data from old buffer to new buffer. The size of meta box and size of exif data is totally unknown, I can not pick a good buffer size for both of them.

Reply:
That really depends if you can pick some reasonable initial size and if you have large amount of data that increasing. For this use case you could just do efree and emalloc instead of realloc as you don't need copying but the point is that sometimes you can re-use the buffer.

If you don't think you can pick a reasonable initial size, you can start with highest limit as zero (containing NULL buffer) and still re-use the buffer and do the efree and emalloc only if needed (if the size is higher than previous highest limit).

Reply:
update code to free and alloc new buffer only if old buffer size is small than exif block